src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java

```diff
@@ -500,6 +500,10 @@ public void buildWebapp(MavenProject project, File webappDirectory)
         for (MavenArtifact artifact : Sets.union(artifacts, dependencyArtifacts)) {
             if (artifact.isPluginBestEffort(getLog()))
                 jenkinsPlugins.add(artifact.getId());
+            // Exclude dependency if it comes from test or provided trail.
+            // Most likely a plugin transitive dependency but the trail through (test,provided) dep is shorter
+            if (artifact.hasScope("test", "provided"))
+                jenkinsPlugins.add(artifact.getId());
```

Comment:
Can you explain what is going on here? `jenkinsPlugins` is supposed to be a list of…well, Jenkins plugins. Which `isPluginBestEffort` is supposed to be detecting. So either that method is broken in some cases (why?); or what you are collecting in this set is not in fact a list of Jenkins plugins but just a list of stuff you want to exclude from processing below, in which case it would be clearer to write something like

```diff
diff --git src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java
index 431052b..d70f0ab 100644
--- src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java
+++ src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java
@@ -508,6 +508,9 @@ public abstract class AbstractHpiMojo extends AbstractJenkinsMojo {
                 continue;   // plugin dependency need not be WEB-INF/lib
             if(artifact.getDependencyTrail().size() >= 1 && jenkinsPlugins.contains(artifact.getDependencyTrail().get(1)))
                 continue;   // no need to have transitive dependencies through plugins in WEB-INF/lib.
+            if (artifact.hasScope("test", "provided")) {
+                continue; // TODO some explanation
+            }
 
             // if the dependency goes through jenkins core, we don't need to bundle it in the war
             // because jenkins-core comes in the <provided> scope, I think this is a bug in Maven that it puts such
```

Reply:
> jenkinsPlugins is supposed to be a list of…well, Jenkins plugins

Well, it's not. It contains all dependencies, regardless their type. Was badly named since the begging. Anyways, I think @Vlatombe has already put it another way in new commits.

Reply:
I separated into a different set for clarity.